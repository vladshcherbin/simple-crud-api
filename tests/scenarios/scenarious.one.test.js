import crypto from 'crypto'
// eslint-disable-next-line import/no-extraneous-dependencies
import { jest } from '@jest/globals'
import request from 'supertest'

const sampleUUID = 'a59ef8b2-a97d-4b17-b31d-2d56fe9d2320'
const mockedRandomUUID = jest.fn(() => crypto.randomUUID())

jest.unstable_mockModule('crypto', () => ({
  default: { ...crypto, randomUUID: mockedRandomUUID }
}))

jest.unstable_mockModule('../../src/data/persons', () => ({
  default: []
}))

const { default: server } = await import('../../src/server')

describe.only('Scenario one', () => {
  test('Return empty person list', async () => {
    const response = await request(server.instance()).get('/person')

    expect(response.status).toBe(200)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Create and return created person', async () => {
    mockedRandomUUID.mockReturnValueOnce(sampleUUID)

    const response = await request(server.instance()).post('/person').send({
      name: 'Bob',
      age: 25,
      hobbies: ['hiking']
    })

    expect(response.status).toBe(201)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Return created person by id', async () => {
    const response = await request(server.instance()).get(`/person/${sampleUUID}`)

    expect(response.status).toBe(200)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Update created person and return updated person', async () => {
    const response = await request(server.instance()).put(`/person/${sampleUUID}`).send({
      name: 'Sam',
      age: 25,
      hobbies: ['singing']
    })

    expect(response.status).toBe(200)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Delete created person', async () => {
    const response = await request(server.instance()).delete(`/person/${sampleUUID}`)

    expect(response.status).toBe(204)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Return 404 error for deleted person', async () => {
    const response = await request(server.instance()).get(`/person/${sampleUUID}`)

    expect(response.status).toBe(404)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })
})
