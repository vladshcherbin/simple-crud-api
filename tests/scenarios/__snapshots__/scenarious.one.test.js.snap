// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario one Create and return created person 1`] = `
Object {
  "age": 25,
  "hobbies": Array [
    "hiking",
  ],
  "id": "a59ef8b2-a97d-4b17-b31d-2d56fe9d2320",
  "name": "Bob",
}
`;

exports[`Scenario one Delete created person 1`] = `""`;

exports[`Scenario one Return 404 error for deleted person 1`] = `
Object {
  "message": "Person not found",
}
`;

exports[`Scenario one Return created person by id 1`] = `
Object {
  "age": 25,
  "hobbies": Array [
    "hiking",
  ],
  "id": "a59ef8b2-a97d-4b17-b31d-2d56fe9d2320",
  "name": "Bob",
}
`;

exports[`Scenario one Return empty person list 1`] = `Array []`;

exports[`Scenario one Update created person and return updated person 1`] = `
Object {
  "age": 25,
  "hobbies": Array [
    "singing",
  ],
  "id": "a59ef8b2-a97d-4b17-b31d-2d56fe9d2320",
  "name": "Sam",
}
`;
